In [6]:
import pandas as pd

In [7]:
X = pd.read_csv('/content/train.csv')
y = pd.read_csv('/content/test.csv')

X.head()

,label,pixel0,pixel1,pixel2,pixel3,pixel4,pixel5,pixel6,pixel7,pixel8,...,pixel774,pixel775,pixel776,pixel777,pixel778,pixel779,pixel780,pixel781,pixel782,pixel783
0,1,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,1,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,4,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [8]:
pixel_cols = [col for col in X.columns if 'pixel' in col]
cnt = 0
for col in pixel_cols:
    unique_vals = X[col].unique()
    print(f"Column: {col}")
    print(f"Unique values: {unique_vals}")

    # Check if all values are between 0 and 255
    is_within_range = (X[col] >= 0).all() and (X[col] <= 255).all()
    if(is_within_range == True): cnt += 1
    print(f"Are all values between 0 and 255?: {is_within_range}")
    print("\n")

print(cnt)

Streaming output truncated to the last 5000 lines.
  97  55 120  79]
Are all values between 0 and 255?: True


Column: pixel439
Unique values: [  0  48 245  96 254 253 255  12 252  20  78 128 161  67  26 189 218   7
 228 206 175 219  14 170 138  62  31  61 242 193 131 231 165 113  60 232
 137 226  47 211 248  10  82  66  23 187 152  59 194   2 247 241  91  38
 246 250 171 160   1 192  76  11  41   9 210 135  22 191 183 188  98  29
 233  94 251 223  53 243  88  33 110 213 149  40  99  55 205   6  74 114
 203   5  27  85 100 136 143  89  56 186 159 197 224  42  35  57 108   8
   3 221 240   4 127  64  13  93  71  32  86 146 244  17  77 179 115  95
  52 119 144 142 150 176 134 120  50 182 209 148 201  24  65  54 125 155
 239  30 102 181 147 163 157  37 207 153 238  44 225 123  21 111 103  25
 195 174  69 140 215 141 168 117 101  19 220  87 116 112 154 145  84 234
  45 118  63  28 196 249 198 124 208 185  15  70 167 104 214 132 227  34
  18 121 126 139 166  46  83 229 129 235 202  16  97 1

In [9]:
# Normalize pixel values from 0-255 to 0-1

# Identify pixel columns (excluding 'label' if it exists in X)
pixel_cols = [col for col in X.columns if 'pixel' in col]

# Normalize training data (X)
X_normalized = X.copy()
for col in pixel_cols:
    X_normalized[col] = X_normalized[col].astype('float32') / 255.0

# Normalize test data (y)
y_normalized = y.copy()
for col in pixel_cols:
    y_normalized[col] = y_normalized[col].astype('float32') / 255.0

# Display the head of the normalized DataFrames
print("Normalized X (Training Data) Head:")
display(X_normalized.head())

print("\nNormalized y (Test Data) Head:")
display(y_normalized.head())

Normalized X (Training Data) Head:


,label,pixel0,pixel1,pixel2,pixel3,pixel4,pixel5,pixel6,pixel7,pixel8,...,pixel774,pixel775,pixel776,pixel777,pixel778,pixel779,pixel780,pixel781,pixel782,pixel783
0,1,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,1,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,4,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0



Normalized y (Test Data) Head:


,pixel0,pixel1,pixel2,pixel3,pixel4,pixel5,pixel6,pixel7,pixel8,pixel9,...,pixel774,pixel775,pixel776,pixel777,pixel778,pixel779,pixel780,pixel781,pixel782,pixel783
0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [10]:
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier

# Separate features (X_data) and labels (y_labels) from the normalized training data
X_data = X_normalized.drop('label', axis=1)
y_labels = X_normalized['label']

# Split the training data into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(X_data, y_labels, test_size=0.2, random_state=42)

models = {
    "tree": DecisionTreeClassifier(),
    "svm_linear": SVC(kernel="linear"),
    "svm_rbf": SVC(kernel="rbf")
}

for name, model in models.items():
    # Fit models on the training set
    model.fit(X_train, y_train)
    # Evaluate models on the validation set (X_val, y_val)
    print(f"{name} validation accuracy: {model.score(X_val, y_val)}")


tree validation accuracy: 0.8555952380952381
svm_linear validation accuracy: 0.9320238095238095
svm_rbf validation accuracy: 0.973452380952381


In [11]:
from sklearn.metrics import confusion_matrix, classification_report

for name, model in models.items():
    print(f"\n--- Evaluation for {name.upper()} Model ---")

    # Make predictions on the validation set
    y_pred = model.predict(X_val)

    # Calculate and print Confusion Matrix
    conf_matrix = confusion_matrix(y_val, y_pred)
    print("Confusion Matrix:\n", conf_matrix)

    # Calculate and print Classification Report
    class_report = classification_report(y_val, y_pred)
    print("Classification Report:\n", class_report)



--- Evaluation for TREE Model ---
Confusion Matrix:
 [[744   2  15   6   6  11   8   6   9   9]
 [  1 865   6   5   4   7   2   4  14   1]
 [  9  19 695  32  16   6  15  23  25   6]
 [  7  14  28 768   9  39   5  14  35  18]
 [  6   9   9   9 708   8  14   7  24  45]
 [  6   8  13  33  11 560  24   7  27  13]
 [  6   5  11  10   9  28 697   0  16   3]
 [ 10   8  22  18   9   2   1 788   7  28]
 [  9  10  30  35  29  24  14   6 658  20]
 [  6   4   6  17  38  21   8  23  11 704]]
Classification Report:
               precision    recall  f1-score   support

           0       0.93      0.91      0.92       816
           1       0.92      0.95      0.93       909
           2       0.83      0.82      0.83       846
           3       0.82      0.82      0.82       937
           4       0.84      0.84      0.84       839
           5       0.79      0.80      0.80       702
           6       0.88      0.89      0.89       785
           7       0.90      0.88      0.89       893
    

In [12]:
print("\n--- Support Vector Analysis ---")

svm_linear_model = models["svm_linear"]
svm_rbf_model = models["svm_rbf"]

print("Linear SVM Support Vectors Shape:", svm_linear_model.support_vectors_.shape)
print("Number of Support Vectors per class (Linear):", svm_linear_model.n_support_)

print("\nRBF SVM Support Vectors Shape:", svm_rbf_model.support_vectors_.shape)
print("Number of Support Vectors per class (RBF):", svm_rbf_model.n_support_)


--- Support Vector Analysis ---
Linear SVM Support Vectors Shape: (6321, 784)
Number of Support Vectors per class (Linear): [370 333 680 820 622 836 441 582 812 825]

RBF SVM Support Vectors Shape: (8469, 784)
Number of Support Vectors per class (RBF): [ 546  388  918  995  916 1085  645  769 1083 1124]


In [13]:
print("\n--- Multiclass Strategy ---")
print("Linear SVM decision function shape:", svm_linear_model.decision_function(X_val[:5]).shape)


--- Multiclass Strategy ---
Linear SVM decision function shape: (5, 10)


In [14]:
from sklearn.decomposition import PCA

print("\n--- PCA Experiment ---")

# Reduce dimensionality
pca = PCA(n_components=50)

X_train_pca = pca.fit_transform(X_train)
X_val_pca = pca.transform(X_val)

# Train SVM again on reduced data
svm_pca = SVC(kernel='rbf')

svm_pca.fit(X_train_pca, y_train)

pca_accuracy = svm_pca.score(X_val_pca, y_val)

print("SVM Accuracy after PCA:", pca_accuracy)


--- PCA Experiment ---
SVM Accuracy after PCA: 0.9777380952380952


PCA reduces dimensionality while preserving most important variance in the data. Even after reducing features from 784 to 50, the SVM model maintains high accuracy, demonstrating that many pixel features are redundant.

| Model         | Accuracy |
| ------------- | -------- |
| Decision Tree | 0.85     |
| Linear SVM    | 0.93     |
| RBF SVM       | 0.97     |
| RBF + PCA     | 0.978    |


The RBF kernel maps the input data into a higher dimensional feature space where nonlinear patterns become linearly separable. Instead of explicitly computing this transformation, the kernel trick allows SVM to compute similarities between data points efficiently.

The nonlinear RBF kernel requires more support vectors to define complex decision boundaries compared to linear SVM.

In [15]:
print("Number of NaN values in test set:", y_normalized.isna().sum().sum())

Number of NaN values in test set: 0


In [16]:
y_normalized = y_normalized[X_data.columns]

In [17]:
# Use the best trained model
best_model = models["svm_rbf"]

# Ensure no NaN values in the test data before prediction
y_normalized = y_normalized.fillna(0)

# Predict digits for Kaggle test data
test_predictions = best_model.predict(y_normalized)

In [18]:
import pandas as pd

submission = pd.DataFrame({
    "ImageId": range(1, len(test_predictions) + 1),
    "Label": test_predictions
})

submission.head()

,ImageId,Label
0,1,2
1,2,0
2,3,9
3,4,9
4,5,3


In [19]:
submission.to_csv("submission.csv", index=False)
print("Submission file saved!")

Submission file saved!
